### PR TITLE
[MODCLUSTER-756] avoid table reads in proxy_cluster_pre_request

### PR DESCRIPTION
**Pull request by aogburn**
https://issues.redhat.com/browse/MODCLUSTER-756

### CODE REVIEW

**Comment by rhusar** [comment]
@jfclere Please review. Also, this will probably need a forward-port PR to 2.x.

File: native/mod_proxy_cluster/mod_proxy_cluster.c [hunk 1]
```diff
@@ -2224,13 +2333,20 @@ static void remove_workers_nodes(proxy_server_conf *conf, apr_pool_t *pool, serv
 static void * APR_THREAD_FUNC proxy_cluster_watchdog_func(apr_thread_t *thd, void *data)
 {
     apr_status_t rv;
-    apr_pool_t *pool;
+    server_rec *s = main_server;
+    void *sconf = s->module_config;
+    proxy_server_conf *conf = (proxy_server_conf *)
+        ap_get_module_config(sconf, &proxy_module);
+    apr_pool_t *cached_pool, *pool;
+    apr_pool_create(&cached_pool, conf->pool);
```
Comment (aogburn) [inline native/mod_proxy_cluster/mod_proxy_cluster.c:2341]: Is this pool suitable for our table allocations?  Or would there be a better means of managing the memory?

File: native/mod_proxy_cluster/mod_proxy_cluster.c [hunk 2]
```diff
@@ -2655,10 +2776,34 @@ static int proxy_cluster_trans(request_rec *r)
     proxy_server_conf *conf = (proxy_server_conf *)
         ap_get_module_config(sconf, &proxy_module);
 
-    proxy_vhost_table *vhost_table = read_vhost_table(r);
-    proxy_context_table *context_table = read_context_table(r);
-    proxy_balancer_table *balancer_table = read_balancer_table(r);
-    proxy_node_table *node_table = read_node_table(r);
+    proxy_vhost_table *vhost_table = cached_vhost_table;
```
Comment (aogburn) [inline native/mod_proxy_cluster/mod_proxy_cluster.c:2779]: Should we lock around this with the table sets in the watchdog thread to ensure we're safely using one set of tables?

**Review by aogburn — COMMENTED** [review]
Some overall questions for consideration in the approaches taken for the performance improvement

File: native/mod_proxy_cluster/mod_proxy_cluster.c (hunk 1 above)
Comment (jfclere) [reply native/mod_proxy_cluster/mod_proxy_cluster.c:2341]: In fact you are leaking the cached_pool each time the proxy_cluster_watchdog_func() is called. That WON'T work.

**Review by jfclere — CHANGES_REQUESTED** [review]
The *_table_cache() and the cached_pool logic leaks memory and needs some locking logic to work correctly.
The cache problem is a lot harder to solve :-(

File: native/mod_proxy_cluster/mod_proxy_cluster.c [hunk 3]
```diff
@@ -2271,8 +2387,13 @@ static void * APR_THREAD_FUNC proxy_cluster_watchdog_func(apr_thread_t *thd, voi
             s = s->next;
         }
         apr_pool_destroy(pool);
-        if (last)
+        if (last) {
```
[the remark this answers is not in the document]
Comment (aogburn) [reply native/mod_proxy_cluster/mod_proxy_cluster.c:2390]: That is likely good enough as a trade off for the gained performance benefit.

File: native/mod_proxy_cluster/mod_proxy_cluster.c [hunk 4]
```diff
@@ -2224,13 +2333,20 @@ static void remove_workers_nodes(proxy_server_conf *conf, apr_pool_t *pool, serv
...
+    cached_vhost_table = read_vhost_table_cached(cached_pool);
+    cached_context_table = read_context_table_cached(cached_pool);
+    cached_balancer_table = read_balancer_table_cached(cached_pool);
+    cached_node_table = read_node_table_cached(cached_pool);
+
     for (;;) {
-        server_rec *s = main_server;
-        void *sconf = s->module_config;
-        proxy_server_conf *conf = (proxy_server_conf *)
-            ap_get_module_config(sconf, &proxy_module);
```
[the remark this answers is not in the document]
Comment (aogburn) [reply native/mod_proxy_cluster/mod_proxy_cluster.c]: Moved it back

File: native/mod_proxy_cluster/mod_proxy_cluster.c (hunk 1 above)
Comment (aogburn) [reply native/mod_proxy_cluster/mod_proxy_cluster.c]: Moved cached pool creation to proxy_cluster_child_init

File: native/mod_proxy_cluster/mod_proxy_cluster.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (aogburn) [reply native/mod_proxy_cluster/mod_proxy_cluster.c:2390]: Should we call apr_pool_destroy on the cached_pool at some point after rereading tables?  Perhaps creating a new cached_pool for the new *_table_cached() reads and keeping a reference to the old cached_pool that is destroyed at some suitable time a bit later (enough to ensure it's not destroyed while some request was still using tables from this pool)?

**Comment by jfclere** [comment]
replaced by https://github.com/modcluster/mod_cluster/pull/617 back ported from upstream

**Pull request closed without merging**